### PR TITLE
fix(50375): Errors for missing enum-named properties should attempt to preserve names

### LINKED ISSUE #50375: Errors for missing enum-named properties should attempt to preserve names
```ts
enum E {
    A
}

let obj: Record<E, any> = {
    // No implementation
}
```

**Expected**:

> Property '[E.A]' is missing in type '{}' but required in type 'Record<E, any>'.

**Actual**:

> Property '0' is missing in type '{}' but required in type 'Record<E, any>'.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #50375

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20310,7 +20310,8 @@ namespace ts {
                     shouldSkipElaboration = true; // Retain top-level error for interface implementing issues, otherwise omit it
                 }
                 if (props.length === 1) {
-                    const propName = symbolToString(unmatchedProperty);
+                    const nameType = getSymbolLinks(unmatchedProperty).nameType;
+                    const propName = nameType && nameType.flags & TypeFlags.EnumLiteral ? `[${symbolToString(nameType.symbol, nameType.symbol.valueDeclaration)}]` : symbolToString(unmatchedProperty);
```
Comment (ProKashif) [inline src/compiler/checker.ts:20314]: We can pass only one argument and then can destructure in symbolToString function right? 

```suggestion
                    const propName = nameType && nameType.flags & TypeFlags.EnumLiteral ? `[${symbolToString(nameType.symbol)}]` : symbolToString(unmatchedProperty);

```

Comment (DanielRosenwasser) [inline src/compiler/checker.ts:20314]: Does it make sense to do this sort of work in `symbolToString` itself?

Comment (jakebailey) [reply src/compiler/checker.ts:20314]: It seems like it... grepping through our errors, we have existing errors like:

```
tests\baselines\reference\computedPropertyNames45_ES5.errors.txt
1:tests/cases/conformance/es6/computedProperties/computedPropertyNames45_ES5.ts(10,5): error TS2411: Property '["get1"]' of type 'Foo' is not assignable to 'string' index type 'Foo2'.
2:tests/cases/conformance/es6/computedProperties/computedPropertyNames45_ES5.ts(11,9): error TS2411: Property '["set1"]' of type 'Foo' is not assignable to 'string' index type 'Foo2'.
17:!!! error TS2411: Property '["get1"]' of type 'Foo' is not assignable to 'string' index type 'Foo2'.
21:!!! error TS2411: Property '["set1"]' of type 'Foo' is not assignable to 'string' index type 'Foo2'.
```

```
tests\baselines\reference\computedPropertyNames38_ES5.errors.txt
1:tests/cases/conformance/es6/computedProperties/computedPropertyNames38_ES5.ts(8,9): error TS2411: Property '[1 << 6]' of type 'Foo' is not assignable to 'string' index type 'Foo2'.
14:!!! error TS2411: Property '[1 << 6]' of type 'Foo' is not assignable to 'string' index type 'Foo2'.
```

So, surely somewhere else we are adding the `[]`, right?

**Review by weswigham — APPROVED** [review]

File: src/compiler/types.ts [hunk 2]
```diff
@@ -4899,6 +4902,9 @@ namespace ts {
         // Prefer aliases which are not directly visible
         UseAliasDefinedOutsideCurrentScope = 0x00000008,
 
+        // { [E.A]: 1 }
+        AllowComputedPropertyEnums = 0x00000009,
```
Comment (jakebailey) [inline src/compiler/types.ts:4906]: I don't think this is right; these flags are powers of two (but not written as shifts for some reason; oversight?), so I don't think 9 is a good value. If the new one is the combination of other flags, then I would write it in terms of them.

**Review by jakebailey — CHANGES_REQUESTED** [review]
This new method seems better, but the flags don't look right. Maybe rewrite them as shifts like the others flags?

File: src/compiler/types.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/compiler/types.ts:4906]: Okay. 
the-simpsons-homer-simpson

Comment (jakebailey) [reply src/compiler/types.ts:4906]: Not your oversight, ours... I truly don't know why this isn't a shift; I asked myself that same question the last time I read this code 😅

File: src/services/codefixes/fixAddMissingMember.ts [hunk 3]
```diff
@@ -642,8 +642,9 @@ namespace ts.codefix {
     }
 
     function createPropertyNameFromSymbol(symbol: Symbol, target: ScriptTarget, quotePreference: QuotePreference, checker: TypeChecker) {
-        if (isTransientSymbol(symbol) && symbol.nameType && symbol.nameType.flags & TypeFlags.UniqueESSymbol) {
-            const expression = checker.symbolToExpression((symbol.nameType as UniqueESSymbolType).symbol, SymbolFlags.Value, symbol.valueDeclaration, NodeBuilderFlags.AllowUniqueESSymbolType);
+        if (isTransientSymbol(symbol) && symbol.nameType && symbol.nameType.flags & (TypeFlags.UniqueESSymbol | TypeFlags.EnumLiteral)) {
+            const nameTypeSymbol = symbol.nameType.symbol;
+            const expression = checker.symbolToExpression(nameTypeSymbol, SymbolFlags.Value, nameTypeSymbol.valueDeclaration, NodeBuilderFlags.AllowUniqueESSymbolType);
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingMember.ts:647]: Doesn't this now need `AllowComputedPropertyEnums`?

**Review by DanielRosenwasser — COMMENTED** [review]
We're running out of flags - does it make sense to have merge any of these flags at all?

**Comment by a-tarasyuk** [comment]
> We're running out of flags - does it make sense to have merge any of these flags at all?

Maybe that makes sense, however, I'm not sure which options can be combined.

**Comment by jakebailey** [comment]
> We're running out of flags - does it make sense to have merge any of these flags at all?

I'm unfortunately not seeing anything we could do to make this better; I _think_ it's the case that we can use up to the full 32-bit range, so it's not a problem now, but...

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 4]
```diff
@@ -2465,15 +2468,17 @@ declare namespace ts {
...
-        NodeBuilderFlagsMask = 848330091
+        NodeBuilderFlagsMask = 1922071915
     }
     export enum SymbolFormatFlags {
         None = 0,
         WriteTypeParametersOrArguments = 1,
         UseOnlyExternalAliasing = 2,
         AllowAnyNodeKind = 4,
         UseAliasDefinedOutsideCurrentScope = 8,
+        WriteComputedProps = 16,
```
Comment (weswigham) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:2481]: Should we mark this new flag as `@internal`, or do we want to make using it part of the public API?

**Comment by DanielRosenwasser** [comment]
> I _think_ it's the case that we can use up to the full 32-bit range, so it's not a problem now, but...

I believe that for 64-bit versions of Node.js that's the case; but 64-bit Electron may have different rules due to its pointer compression encoding (see https://github.com/microsoft/vscode/issues/151245). [If I've read this V8 blog post correctly](https://v8.dev/blog/pointer-compression), under that mode an SMI doesn't get a full 32 bits.

File: src/compiler/types.ts [hunk 5]
```diff
@@ -4545,6 +4545,8 @@ namespace ts {
         /** Note that the resulting nodes cannot be checked. */
         symbolToExpression(symbol: Symbol, meaning: SymbolFlags, enclosingDeclaration: Node | undefined, flags: NodeBuilderFlags | undefined): Expression | undefined;
         /** Note that the resulting nodes cannot be checked. */
+        symbolToNode(symbol: Symbol, meaning: SymbolFlags, enclosingDeclaration: Node | undefined, flags: NodeBuilderFlags | undefined): Node | undefined;
```
Comment (sheetalkamat) [inline src/compiler/types.ts:4548]:
```suggestion
        /*@internal*/ symbolToNode(symbol: Symbol, meaning: SymbolFlags, enclosingDeclaration: Node | undefined, flags: NodeBuilderFlags | undefined): Node | undefined;
```

**Comment by jakebailey** [comment]
If Electron behaves differently, maybe we should be testing it; I bet it'd be straightforward to create a builder that does an `ELECTRON_RUN_AS_NODE=1` or something (and even use `@vscode/test-electron` to grab the latest VS Code and run it).

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**